File: dio-codes/src/intermediate/cofres_seguros/CofresSeguros.kt
package intermediate.cofres_seguros

import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private fun getInput(): Array<String> {
    val reader = BufferedReader(InputStreamReader(System.`in`), 30)
    val input = arrayOf(reader.readLine(), reader.readLine(), reader.readLine())
    reader.close()
    return input
}

private fun interface Vault {
    companion object {
        fun instanceOf(tipo: String): Vault = if (tipo == "digital") VaultDigital() else VaultPhysic()
    }

    fun open(type: String, pass: String, confirmPass: String): String
}

private class VaultDigital : Vault {
    override fun open(type: String, pass: String, confirmPass: String): String =
        if (validatePass(pass, confirmPass)) "Tipo: Cofre Digital\r\nMetodo de abertura: Senha\r\nCofre aberto!"
        else "Tipo: Cofre Digital\r\nMetodo de abertura: Senha\r\nSenha incorreta!"

    private fun validatePass(pass: String, confirmPass: String): Boolean = pass == confirmPass
}

private class VaultPhysic : Vault {
    override fun open(type: String, pass: String, confirmPass: String): String =
        "Tipo: Cofre Fisico\r\nMetodo de abertura: Chave"
}

fun main() {
    val input = getInput()
    val vault: Vault = Vault.instanceOf(input[0])
    val output: CharSequence = vault.open(input[0], input[1], input[2])
    OutputStreamWriter(System.out).append(output).close()
}

File: dio-codes/src/intermediate/cofres_seguros/cofresSeguros.ts
import reader from "readline-sync"

abstract class Vault {
    static instanceOf = (type: string): Vault => (type === "digital") ? new VaultDigital() : new VaultPhysic()

    abstract open(type: string, pass: string, confirmPass: string): string;
}

class VaultDigital implements Vault {
    open = (_type: string, pass: string, confirmPass: string) => this.validatePass(pass, confirmPass) ? "Tipo: Cofre Digital\r\nMetodo de abertura: Senha\r\nCofre aberto!" : "Tipo: Cofre Digital\r\nMetodo de abertura: Senha\r\nSenha incorreta!"
    private validatePass = (pass: string, confirmPass: string) => pass === confirmPass
}

class VaultPhysic implements Vault {
    open = (_tipo: string) => "Tipo: Cofre Fisico\r\nMetodo de abertura: Chave"
}

function main() {
    let input = [reader.question(), reader.question(), reader.question()]
    let vault: Vault = Vault.instanceOf(input[0])
    let output = vault.open(input[0], input[1], input[2])
    console.log(output)
}

main()